File: solutions/0020. Valid Parentheses/0020.js
// Time complexity: O(n)
// Space complexity: O(n)
var isValid = function(s) {
    const st = [];

    for (let c of s) {
        if (c === '(') {
            st.push(')');
        } else if (c === '{') {
            st.push('}');
        } else if (c === '[') {
            st.push(']');
        } else if (!st.length || st.pop() !== c) {
            return false;
        }
    }
    return !st.length;
}

File: solutions/0020. Valid Parentheses/0020.cs
// Time complexity: O(n)
// Space complexity: O(n)
public class Solution {
    public bool IsValid(string s) {
        Stack<char> st = new Stack<char>();

        foreach(char c in s) {
            if (c == '(') {
                st.Push(')');
            } else if (c == '{') {
                st.Push('}');
            } else if (c == '[') {
                st.Push(']');
            } else {
                if (st.Count == 0 || c != st.Pop()) {
                    return false;
                }
            }
        }
        return st.Count == 0;
    }
}